examples/optimize_evoked.py

```diff
@@ -0,0 +1,81 @@
+"""
+=================================================
+08. Optimize simulated evoked response parameters
+=================================================
+
+This example demonstrates how to optimize the parameters
+of the model simulation to match an experimental dipole waveform.
+"""
+
+# Authors: Blake Caldwell <blake_caldwell@brown.edu>
+#          Mainak Jas <mjas@mgh.harvard.edu>
+
+import os.path as op
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+###############################################################################
+# Let us import hnn_core
+
+import hnn_core
+from hnn_core.dipole import simulate_dipole
+from hnn_core import read_params, Dipole, MPIBackend, Network
+
+hnn_core_root = op.join(op.dirname(hnn_core.__file__))
+
+n_procs = 10
+
+###############################################################################
+# First, we will load experimental data into Dipole object.
+#
+# This is a different experiment than the one to which the base parameters were
+# tuned. So, the initial RMSE will be large, giving the optimization procedure
+# a lot to work with.
+from urllib.request import urlretrieve
+
+data_url = ('https://raw.githubusercontent.com/jonescompneurolab/hnn/master/'
+            'data/MEG_detection_data/S1_SupraT.txt')
+urlretrieve(data_url, 'S1_SupraT.txt')
+extdata = np.loadtxt('S1_SupraT.txt')
+exp_dpl = Dipole(extdata[:, 0], np.c_[extdata[:, 1],
+                 extdata[:, 1], extdata[:, 1]])
+
+###############################################################################
+# Read the base parameters from a file
+params_fname = op.join(hnn_core_root, 'param', 'default.json')
+params = read_params(params_fname)
+
+###############################################################################
+# Let's first simulate the dipole with the initial parameters.
+net = Network(params, add_drives_from_params=True)
+with MPIBackend(n_procs=n_procs):
+    initial_dpl = simulate_dipole(net, n_trials=1)
+
+###############################################################################
+# Start the optimization!
+
+from hnn_core.optimization import optimize_evoked
+
+with MPIBackend(n_procs=n_procs):
+    params_optim = optimize_evoked(params, exp_dpl)
```

Comment:
In the spirit of moving away from working with the `params` dictionary correctly, perhaps you could pass a copy of the `Network` object that gets modified in place?

Reply:
This is tricky because if you're going to optimized a bunch of parameters, it's nice to be able to find them all in one place in order to see what changed.

Reply:
Even though I agree in principle with @ntolley I'd rather defer this to a future PR